DC/OS HTTP Transport

malnick opened the pull request:
Adds generic HTTP client code to dcos-go. 

- [x] ~80% unit test coverage?
- [x] `cmd/example/example_usage.go` present?
- [x] Updated, local README.md describing package code?

malnick commented:
@jdef Can you give me a quick review when you get a minute? @darkonie is on vaca.

darkonie commented on dcos/http/client/client.go:
```diff
@@ -0,0 +1,93 @@
+// loadCAPool will load a valid x509 cert.
+func loadCAPool(path string) (*x509.CertPool, error) {
+	caPool := x509.NewCertPool()
+	f, err := os.Open(path)
```
can we use `ioutil.ReadFile()` instead?

malnick replied:
+1, appears os.Open, though used by ioutil.ReadFile, does not implement an implicit .Close() https://golang.org/src/io/ioutil/ioutil.go?s=1464:1510#L39

darkonie commented on dcos/http/client/client.go:
```diff
@@ -0,0 +1,93 @@
+// a Authorization header when passed a IAM configuration path. If the path to CA
+// certificate or IAM config is an empty string then the corrosponding client
+// configurations which they set are ignored.
+func NewDCOSClient(caCertificatePath string, iamConfigPath string) (*http.Client, error) {
```
maybe instead of returning `*http.Client` we should return `http.RoundTripper`?
so we can make a roundtripper decorator similar to https://github.com/dcos/dcos-go/blob/master/jwt/transport/transport.go#L48 and implement the IAM Config as a functional parameter. We could also refactor `jwt/transport` to include this functionality with a new option? So we could handle JWT and CA certs all from one place.

i feel like `http.Client` constructor is too broad.
@jdef what's your opinion?

malnick replied:
If we use a RoundTripper then the caller has to construct their own client, which defeats the purpose of this helper method, which is trying to abstract all the DC/OS specific client code away. 

If we do decide that a DC/OS client is to generic for this  library, I'd vote to add the CAPool loading to the JWT round tripper code. In that event, I'd like to rename the jwt/transport package to something like dcos/http/roundtripper to be more clear about the purpose of the round tripper implementation.

jdef replied:
you could split the baby:
```go
// NewTransport creates and returns an HTTP configured with the given certs and (optional) IAM configuration
func NewTransport(caCertificatePath string, iamConfigPath string) (http.RoundTripper, error) {
  // most of the code from the existing NewDCOSClient (just don't create an http.Client)
  tr, err := newTransport(caCertificatePath)
  if err != nil {
	return nil, err
  }
  if len(iamConfigPath) != 0 {
    withIAM, err := transport.NewRoundTripper(
      tr,
      transport.OptionReadIAMConfig(iamConfigPath),
    )
    if err != nil {
      return nil, err
    }
    return withIAM
  }
  return tr
}

// NewClient creates and returns an HTTP client (based on http.DefaultClient), configured with an HTTP transport via NewTransport
func NewClient(caCertificatePath string, iamConfigPath string) (*http.Client, error) {
  tr, err := NewTransport(caCertificatePath, iamConfigPath)
  if err != nil {
    return nil, err
  }
  client := *http.DefaultClient
  client.Transport = tr
  return &client, nil
}
```

you might even consider refactoring `newTransport` to accept an optional (nil-able) `*http.Transport` param. if it's nil, then use `&http.Transport{}` like you do now. once that's in place, you could add a similar optional (nil-able) param to `NewTransport` so that a caller could provide the base transport layer.

(notice s/NewDCOSClient/NewClient/ to reduce stutter)

jdef commented on dcos/http/client/client.go:
```diff
@@ -0,0 +1,93 @@
+}
+
+// getTransport will return transport for http.Client
+func getTransport(caCertificatePath string) (*http.Transport, error) {
```
nit: s/getTransport/newTransport/

jdef commented on dcos/http/client/client.go:
```diff
@@ -0,0 +1,93 @@
+			client.Transport,
+			transport.OptionReadIAMConfig(iamConfigPath))
+		if err != nil {
+			return client, err
```
why return `client` instead of `nil` upon error?

jdef commented on dcos/http/client/client_test.go:
```diff
@@ -0,0 +1,76 @@
+		t.Error("Expected nil error getting transport, got", err.Error())
+	}
+
+	if reflect.TypeOf(transport) != reflect.TypeOf(&http.Transport{}) {
```
not sure how useful this as a "test". the interface for `getTransport` declares that it returns `*http.Transport` -- why bother testing it? it can't return anything else...

jdef commented on dcos/http/client/client_test.go:
```diff
@@ -0,0 +1,76 @@
+		t.Error("Expected nil error, got", err.Error())
+	}
+
+	if reflect.TypeOf(client) != reflect.TypeOf(&http.Client{}) {
```
ditto. the type returned here is a struct pointer. what's the value in testing that?

jdef commented on dcos/http/client/client_test.go:
```diff
@@ -0,0 +1,76 @@
+		t.Error("Expected nil error, got", err.Error())
+	}
+
+	if reflect.TypeOf(caClient) != reflect.TypeOf(&http.Client{}) {
```
ditto

jdef commented on dcos/http/client/cmd/example/main.go:
```diff
@@ -0,0 +1,58 @@
+		panic(err)
+	}
+
+	req, _ := http.NewRequest("GET", *url, nil)
```
why bother with the mechanics of declaring a request and calling `Do`. why not `client.Get` instead?

jdef commented on dcos/http/client/cmd/example/main.go:
```diff
@@ -0,0 +1,58 @@
+			log.Fatal(err)
+		}
+
+		fmt.Println(string(body))
```
alternative approach:
```
_, err = io.Copy(os.Stdout, body)
if err != nil {
  log.Fatal(err)
}
```
... if you don't want to buffer the whole thing as a string just to re-print it

jdef commented on dcos/http/client/fixtures/root_ca.key:
```diff
@@ -0,0 +1,27 @@
+-----BEGIN RSA PRIVATE KEY-----
```
where is this file actually referenced?

malnick replied:
It's not but I used it to make the CA cert.

malnick commented:
@jdef @darkonie Updated with recommendations.

darkonie commented on dcos/http/client/client.go:
```diff
@@ -0,0 +1,33 @@
+// a Authorization header when passed a IAM configuration path. If the path to CA
+// certificate or IAM config is an empty string then the corrosponding client
+// configurations which they set are ignored.
+func NewClient(caCertificatePath string, iamConfigPath string) (*http.Client, error) {
```
can we make it to accept functional parameters here? i assume we may want to extend NewClient in future with DC/OS specific options. This will ensure interface isn't broken for current users.

malnick replied:
+1

malnick commented:
@darkonie this is ready for another reivew:
- Removed old jwt/transport lib
- Added jwt/transport to roundtripper.go and roundtripper_test.go
- Made new cmd examples from jwt/transport cmd examples under cmd/examples/roundtripper_example

Tests pass, and I think this new organization makes a lot of sense. Let me know your thoughts.

malnick commented on dcos/http/transport/roundtripper.go:
```diff
@@ -26,31 +26,30 @@ var (
 	ErrEmptyToken = errors.New("Empty token")
```
Need to add (C) to top of file.

malnick commented on dcos/http/transport/roundtripper_test.go:
```diff
@@ -74,14 +74,14 @@ func TestNewRoundTripper(t *testing.T) {
 		},
```
Needs copywrite (c) at top of file

malnick commented on dcos/http/transport/cmd/example/transport_example/transport_example.go:
```diff
@@ -0,0 +1,59 @@
+package main
```
Needs copywrite (c) at top of file

darkonie commented on dcos/http/transport/transport.go:
```diff
@@ -0,0 +1,96 @@
+func NewTransport(clientOptionFuncs ...OptionFunc) (http.RoundTripper, error) {
+	transportOptions := options{}
+	for _, opt := range clientOptionFuncs {
+		if err := opt(&transportOptions); err != nil {
```
check for nil needed here.
nil is a valid OptionFunc

darkonie commented on dcos/http/transport/transport.go:
```diff
@@ -0,0 +1,96 @@
+		return withIAM, nil
+	}
+
+	return tr, nil
```
what if instead of returning `http.Transport{}` we will return our implementation of `http.RoundTripper`? i think this will be much more extendible in future.

malnick replied:
You can get a DC/OS roundtripper by calling the new method directly. You can also get a DC/OS transport by calling its New() method too. The transport takes care of the complete TLS picture while the roundtripper takes care of the JWT picture.

malnick replied:
Ugh, never mind. I see what I did. I need to merge your roundtripper into my transport and have a single roundtripper.go that does all of it. Right now we basically have two implementations of a transport since a net/http/transport implements roundtripper. 

Third time is a charm I hope.

darkonie commented on dcos/http/transport/option.go:
```diff
@@ -0,0 +1,121 @@
+
+func errorOnEmpty(arg string) error {
+	if len(arg) == 0 {
+		return errors.New("Must pass non-empty string to this option")
```
can we extract the error and make it package global, similar to `ErrInvalidCredentials` and `ErrInvalidExpireDuration`

darkonie commented:
🚢
